lib/dialects/abstract/connection-manager.js

```diff
@@ -248,56 +217,65 @@ class ConnectionManager {
-          const _options = {};
-          _options.transaction = {connection}; // Cheat .query to use our private connection
-          _options.logging = () => {};
-          _options.logging.__testLoggingFn = true;
-
-          return this.sequelize.databaseVersion(_options).then(version => {
-            this.sequelize.options.databaseVersion = semver.valid(version) ? version : this.defaultVersion;
-            this.versionPromise = null;
+        promise = this.versionPromise = this._connect(this.config.replication.write || this.config)
+          .then(mayBeConnection => this._determineConnection(mayBeConnection))
+          .then(connection => {
+            const _options = {};
+
+            _options.transaction = {connection}; // Cheat .query to use our private connection
+            _options.logging = () => {};
+            _options.logging.__testLoggingFn = true;
+
+            return this.sequelize.databaseVersion(_options).then(version => {
+              this.sequelize.options.databaseVersion = semver.valid(version) ? version : this.defaultVersion;
+              this.versionPromise = null;
 
-            return this._disconnect(connection);
+              return this._disconnect(connection);
+            });
+          }).catch(err => {
+            this.versionPromise = null;
+            throw err;
           });
-        }).catch(err => {
-          this.versionPromise = null;
-          throw err;
-        });
       }
     } else {
       promise = Promise.resolve();
     }
 
     return promise.then(() => {
-      return Promise.race([
-        this.pool.acquire(options.priority, options.type, options.useMaster),
-        new Promise((resolve, reject) =>
-          timers.setTimeout(() => {
-            if (this.poolError) {
-              reject(this.poolError);
-            }
-          }, 0))
-      ])
-        .tap(() => { debug('connection acquired'); })
-        .catch(e => {
-          e = this.poolError || e;
-          this.poolError = null;
-          throw e;
-        });
+      return this.pool.acquire(options.priority, options.type, options.useMaster)
+        .then(mayBeConnection => this._determineConnection(mayBeConnection))
+        .tap(() => { debug('connection acquired'); });
     });
   }
 
   releaseConnection(connection) {
-    return this.pool.release(connection).tap(() => {
-      debug('connection released');
+    return this.pool.release(connection)
+      .tap(() => { debug('connection released'); })
+      .catch(/Resource not currently part of this pool/, () => {});
+  }
+
+  _determineConnection(mayBeConnection) {
+    return new Promise((resolve, reject) => {
+      if (mayBeConnection instanceof Error) {
+        return this.pool.destroy(mayBeConnection)
+          .then(() => { reject(mayBeConnection); })
+          .catch(/Resource not currently part of this pool/, () => {
+            reject(mayBeConnection);
+          });
+      }
+
+      return resolve(mayBeConnection);
     });
```

Comment:
How do you feel about this? Since you're using bluebird, it reduces a little boilerplate and I think this is equivalent.

```javascript
_determineConnection(mayBeConnection) {
  if (mayBeConnection instanceof Error) {
    return Promise.resolve(this.pool.destroy(mayBeConnection))
      .catch(/Resource not currently part of this pool/, () => {}) // suppress this error
      .then(() => { throw mayBeConnection; });
  }

  return Promise.resolve(mayBeConnection);
}
```